### PR TITLE
Reverse library linking order

### LINKED ISSUE #1828: Linking order confusion for libraries
`block_bind_library` (and `block_bind_each`, `block_bind_referenced`, and maybe others) traverse the binder block from first by next, so the first function declared with a given name gets used, even if the library contains another function of the same name declared later.

The parser, on the other hand, binds backwards, and later definitions "override" (really, get locked in before) those earlier in the file. This leads to an inconsistency:

```
% jq -n 'def a: 1; def a: 2; a'
2
% echo 'def a: 1; def a: 2;' > a.jq
% jq -n 'include "a"; a'
1
```

This is a pretty trifling point, but I'm leaving it here for consideration because from cursory examination, the change I suggested in https://github.com/stedolan/jq/issues/1826#issuecomment-464597334 would require doing `bind_block_referenced`s _backward_ in order to get the linkage within builtin.jq right, which seems like a good opportunity to do the same thing when linking libraries, which would (from my perspective) fix this inconsistency.

There's a bit of trickiness when a single file includes multiple libraries; the libraries need to be bound in backwards, too, I believe.

Thoughts?

### PR DESCRIPTION
**Pull request by muhmuhten**
Actually fixes #1828.

A breaking change, as noted, and changes the test_bind_order test, but on the bright side it only affects situations where exported function names are shadowed, which is a bad situation to be in in the first place.

### CODE REVIEW

**Comment by nicowilliams** [comment]
Is this still relevant?

**Comment by muhmuhten** [comment]
Yes, we're still binding the first definition in a library instead of the last one.

**Comment by nicowilliams** [comment]
@muhmuhten can you rebase and merge the conflicts?

**Comment by muhmuhten** [comment]
@nicowilliams got it.

**Comment by nicowilliams** [comment]
I had started, but got interrupted before i could finish it, otherwise I'd have saved you the bother.

File: src/linker.c [hunk 1]
```diff
@@ -239,7 +240,11 @@ static int process_dependencies(jq_state *jq, jv jq_origin, jv lib_origin, block
   block bk = *src_block;
   int nerrors = 0;
 
-  jv_array_foreach(deps, i, dep) {
+  // XXX This is a backward jv_array_foreach because bindings go in reverse
+  for (int i = jv_array_length(jv_copy(deps)); i > 0; ) {
```
Comment (nicowilliams) [inline src/linker.c]: a) Maybe we ought to have a macro for this too,

b) I've long wanted to do something a wee bit safer w.r.t. lengths.  A signed length works well for a backwards foreach, but for just about everything else I'd rather use `size_t` due to it being unsigned (which has well-defined overflow semantics in C).

I never convinced @stedolan of (b) and I'm not going to wait for (a).

File: tests/jq.test [hunk 2]
```diff
@@ -604,6 +604,10 @@ def f(a;b;c;d;e;f): [a+1,b,c,d,e,f]; f(.[0];.[1];.[0];.[0];.[0];.[0])
 [1,2]
 [2,2,1,1,1,1]
 
+def f: 1; def g: f, def f: 2; def g: 3; f, def f: g; f, g; def f: 4; [f, def f: g; def g: 5; f, g]+[f,g]
```
Comment (nicowilliams) [inline tests/jq.test]: Oof, this one is hard to parse.  For anyone following along:

```
def f: 1;
def g: f,
       def f: 2;
       def g: 3;
       f,
       def f: g;
       f, g;
def f: 4;
[f, def f: g; def g: 5; f, g]+[f,g]
```
The first `f` in `[f, ...]` will be `4`, the second will be `g`, which will b `1, 2, 3, 3`, then `g` will be `5`.  The we add `[f,g]` to that, which will be the first `f` and `g`, so, again `[1,2,3,3]`.

**Comment by nicowilliams** [comment]
Thanks!
